yt_dlp/extractor/youtube.py

```diff
@@ -2643,30 +2643,53 @@ def _extract_signature_timestamp(self, video_id, player_url, ytcfg=None, fatal=F
         return sts
 
     def _mark_watched(self, video_id, player_responses):
-        playback_url = get_first(
-            player_responses, ('playbackTracking', 'videostatsPlaybackUrl', 'baseUrl'),
-            expected_type=url_or_none)
-        if not playback_url:
+        mark_watched_urls = {
+            "playback": get_first(
+                player_responses, ('playbackTracking', 'videostatsPlaybackUrl', 'baseUrl'),
+                expected_type=url_or_none),
+            "watchtime": get_first(
+                player_responses, ('playbackTracking', 'videostatsWatchtimeUrl', 'baseUrl'),
+                expected_type=url_or_none)
+        }
+
+        if not all(mark_watched_urls.values()):
             self.report_warning('Unable to mark watched')
             return
-        parsed_playback_url = compat_urlparse.urlparse(playback_url)
-        qs = compat_urlparse.parse_qs(parsed_playback_url.query)
-
-        # cpn generation algorithm is reverse engineered from base.js.
-        # In fact it works even with dummy cpn.
-        CPN_ALPHABET = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_'
-        cpn = ''.join(CPN_ALPHABET[random.randint(0, 256) & 63] for _ in range(0, 16))
-
-        qs.update({
-            'ver': ['2'],
-            'cpn': [cpn],
-        })
-        playback_url = compat_urlparse.urlunparse(
-            parsed_playback_url._replace(query=compat_urllib_parse_urlencode(qs, True)))
-
-        self._download_webpage(
-            playback_url, video_id, 'Marking watched',
-            'Unable to mark watched', fatal=False)
+        for url_type, url in mark_watched_urls.items():
+            parsed_url = compat_urlparse.urlparse(url)
+            qs = compat_urlparse.parse_qs(parsed_url.query)
+
+            # cpn generation algorithm is reverse engineered from base.js.
+            # In fact it works even with dummy cpn.
+            CPN_ALPHABET = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_'
+            cpn = ''.join(CPN_ALPHABET[random.randint(0, 256) & 63] for _ in range(0, 16))
+
+            video_length = qs.get('len') or ['1.5']
+            # more consistent results setting it to right before the end
+            video_length_float = float(video_length[0]) - 1
+            video_length = [str(video_length_float)]
```

Comment:
`qs.get('len', ['1.5'])` and `qs.get('len') or ['1.5']` are semantically different, because the latter guards against the case that the `len` query param exists but is blank (which would crash if you try to pass it into `float`)

Reply:
Nope, which is why I recomended that. But either is fine tbh
```py
>>> from urllib.parse import parse_qs
>>> parse_qs('a=1&b')
{'a': ['1']}
>>> parse_qs('a=1&b=')
{'a': ['1']}
>>> parse_qs('a=1&b=&c=1')
{'a': ['1'], 'c': ['1']}
```